In [1]:
import numpy as np
import pandas as pd
import os

In [2]:
import warnings
warnings.filterwarnings("ignore")

# Read raw data

In [3]:
dir_main = os.getcwd()
dir_raw_data = os.path.join(dir_main, "dataset", "raw")

## S&P500

In [4]:
sp = pd.read_csv(os.path.join(dir_raw_data, "sp500.csv"),
                sep=",",
                header=0,
                dtype={"Date": object, "Close/Last": float,
                       "Open": float, "High": float, "Low": float})

In [5]:
# Rename columns for readability
sp.rename(columns={"Date": 'date', "Close/Last": 'close',
                   "Open": 'open', "High": 'high', "Low": 'low'},
                   inplace=True)

# Set date column's data type
sp['date'] = pd.to_datetime(sp['date'])

# Sort rows by ascending order of date
sp = sp.sort_values('date')

In [6]:
# Count null values
sp.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2538 entries, 2537 to 0
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   date    2538 non-null   datetime64[ns]
 1   close   2538 non-null   float64       
 2   open    2538 non-null   float64       
 3   high    2538 non-null   float64       
 4   low     2538 non-null   float64       
dtypes: datetime64[ns](1), float64(4)
memory usage: 119.0 KB


In [7]:
# Confirm first and last date
print("First date:\t", sp['date'].iloc[0], sep="")
print("Latest date:\t", sp['date'].iloc[-1], sep="")

First date:	2013-11-11 00:00:00
Latest date:	2023-11-10 00:00:00


## Effective Federal Reserve Fund Rate

In [8]:
effr = pd.read_csv(os.path.join(dir_raw_data, "EFFR.csv"),
                   sep=",",
                   header=0,
                   dtype={"DATE": object, "EFFR": float},
                   na_values=["."]
                   )

In [9]:
# Rename columns for readability
effr.rename(columns={"DATE": 'date', "EFFR": 'effr'}, inplace=True)

# Set date column's data type
effr['date'] = pd.to_datetime(effr['date'])

# Sort rows by ascending order of date
effr = effr.sort_values('date')

In [10]:
# Count null values
effr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2608 entries, 0 to 2607
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   date    2608 non-null   datetime64[ns]
 1   effr    2512 non-null   float64       
dtypes: datetime64[ns](1), float64(1)
memory usage: 40.9 KB


In [11]:
# Get list of null dates
null_dates = effr['date'][effr['effr'].isna()]

In [12]:
# Confirm first and last date
print("First date:\t", effr['date'].iloc[0], sep="")
print("Latest date:\t", effr['date'].iloc[-1], sep="")

First date:	2013-11-12 00:00:00
Latest date:	2023-11-09 00:00:00


## US GDP

In [13]:
gdp = pd.read_csv(os.path.join(dir_raw_data, "GDP.csv"),
                  sep=",",
                  header=0,
                  dtype={"DATE": object, "GDP": float})

In [14]:
# Rename columns for readability
gdp.rename(columns={"DATE": "date", "GDP": "gdp"}, inplace=True)

# Set date column's data type
gdp['date'] = pd.to_datetime(gdp['date'])

# Sort rows by ascending order of date
gdp = gdp.sort_values('date')

In [15]:
# Count null values
gdp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40 entries, 0 to 39
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   date    40 non-null     datetime64[ns]
 1   gdp     40 non-null     float64       
dtypes: datetime64[ns](1), float64(1)
memory usage: 768.0 bytes


In [16]:
# Confirm first and last date
print("First date:\t", gdp['date'].iloc[0], sep="")
print("Latest date:\t", gdp['date'].iloc[-1], sep="")

First date:	2013-10-01 00:00:00
Latest date:	2023-07-01 00:00:00


## MSCI ACWI ETF

In [17]:
acwi = pd.read_csv(os.path.join(dir_raw_data, "ACWI.csv"),
                  sep=",",
                  header=0,
                  dtype={"Date": object, "Open": float, "High": float,
                         "Low": float, "Close": float,
                         "Adj Close": float, "Volume": float})

In [18]:
# Remove unnecessary columns
acwi = acwi[['Date', 'Open']]

# Rename columns for readability
acwi.rename(columns={"Date": "date", "Open": "acwi_open"}, inplace=True)

# Set date column's data type
acwi['date'] = pd.to_datetime(acwi['date'])

# Sort rows by ascending order of date
acwi = acwi.sort_values('date')

In [19]:
acwi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2547 entries, 0 to 2546
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   date       2547 non-null   datetime64[ns]
 1   acwi_open  2547 non-null   float64       
dtypes: datetime64[ns](1), float64(1)
memory usage: 39.9 KB


In [20]:
# Confirm first and last date
print("First date:\t", acwi['date'].iloc[0], sep="")
print("Latest date:\t", acwi['date'].iloc[-1], sep="")

First date:	2013-10-01 00:00:00
Latest date:	2023-11-10 00:00:00


## Dataset summary

#### S&P500 Dataset
* 2538 rows
* From `2013-11-11` to `2023-11-10`
* `Daily` basis
* Schema: `['date', 'close', 'open', 'high', 'low']`
* no null value

#### Effective Federal Reserve Fund Rate
* 2608 rows
* From `2013-11-12` to `2023-11-09`
* `Daily` basis
* Schema: `['date', 'effr']`
* effr has 96 null values for holidays or other days.
    * These dates will be removed from every datasets.

#### US GDP
* 40 rows
* From `2013-10-01` to `2023-07-01`
* `Quarterly` basis
    * Latest quarterly GDP value will be used for each given date.
* Schema: `['date', 'gdp']`
* no null value

#### MSCI ACWI ETF
* 2547 rows
* From `2013-10-01` to `2023-11-10`
* `Daily` basis
* Schema: `['date', 'acwi_open']`
* no null value

# Combine(Join) Dataset

## Schema of cleaned data
`['date', 'close', 'open', 'high', 'low', 'effr', 'gdp', 'acwi_open']`
* `date`: every date in `S&P500`, but from `2013-11-12` to `2023-11-09`
    * The earliest and latest dates are bounded by EFFR data
* `close`, `open`, `high`, `low`: data from `S&P500`
* `effr`: data from `Effective Federal Reserve Fund Rate`
* `gdp`: data from `US GDP`
* `acwi_open`: Opening price from `Open` in `MSCI ACWI`

In [21]:
# Generate valid date pd.series
valid_date = sp['date']
mask_validDate = (valid_date >= pd.to_datetime('2013-11-12')) &\
                 (valid_date <= pd.to_datetime('2023-11-09')) &\
                 (~valid_date.isin(null_dates))
valid_date = valid_date[mask_validDate]

In [22]:
# Change US GDP from quarterly to daily values


In [23]:
# Merge with S&P500
dataset = pd.merge(valid_date, sp, on='date', how='inner')

# Merge with US GDP; apply latest quarterly GDP at any given date
dataset = pd.merge_asof(dataset, gdp, on='date', direction='backward')

# Merge with Effective Federal Fund Rate
dataset = pd.merge(dataset, effr, on='date', how='inner')

# Merge with MSCI ACWI ETF
dataset = pd.merge(dataset, acwi, on='date', how='inner')

### Normalize each column

In [24]:
from sklearn.preprocessing import StandardScaler

In [25]:
# Standardization for every column except date
scaler = StandardScaler()
columns_to_standardize = dataset.columns.drop('date')
dataset[columns_to_standardize] = scaler.fit_transform(dataset[columns_to_standardize])

In [26]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2498 entries, 0 to 2497
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   date       2498 non-null   datetime64[ns]
 1   close      2498 non-null   float64       
 2   open       2498 non-null   float64       
 3   high       2498 non-null   float64       
 4   low        2498 non-null   float64       
 5   gdp        2498 non-null   float64       
 6   effr       2498 non-null   float64       
 7   acwi_open  2498 non-null   float64       
dtypes: datetime64[ns](1), float64(7)
memory usage: 156.2 KB


## Save/Read Dataset

In [27]:
dir_main = os.getcwd()
dir_prepared_data = os.path.join(dir_main, "dataset", "prepared")

### Save dataset

In [28]:
dataset.to_csv(os.path.join(dir_prepared_data, "dataset.csv"),
               index=False)

### Read dataset

In [29]:
dataset = pd.read_csv(os.path.join(dir_prepared_data, "dataset.csv"),
                      sep=",",
                      header=0,
                      dtype={"date": object, "close": float, "open": float,
                             "high": float, "low": float,
                             "gdp": float, "effr": float,
                             "acwi_open": float})
dataset['date'] = pd.to_datetime(dataset['date'])
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2498 entries, 0 to 2497
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   date       2498 non-null   datetime64[ns]
 1   close      2498 non-null   float64       
 2   open       2498 non-null   float64       
 3   high       2498 non-null   float64       
 4   low        2498 non-null   float64       
 5   gdp        2498 non-null   float64       
 6   effr       2498 non-null   float64       
 7   acwi_open  2498 non-null   float64       
dtypes: datetime64[ns](1), float64(7)
memory usage: 156.2 KB
